File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int numberOfIslands = 0;
        int r = grid.size();
        int c = grid[0].size();
        unordered_set<string> visited;

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (grid[i][j] == '1' &&
                    visited.find((to_string(i) + "," + to_string(j))) ==
                        visited.end()) {
                    numberOfIslands++;
                    bfs(grid, visited, i, j);
                }
            }
        }
        return numberOfIslands;
    }

private:
    vector<pair<int, int>> dir = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
    
    void bfs(vector<vector<char>>& grid, unordered_set<string>& visited, int r,
             int c) {
        string coor = to_string(r) + "," + to_string(c);
        visited.insert(coor);
        queue<pair<int, int>> q;
        q.push({r, c});

        while (!q.empty()) {
            auto [row, col] = q.front();
            q.pop();
            for (auto [l, r] : dir) {
                int rr = row + l;
                int cc = col + r;

                if (rr >= 0 && rr < grid.size() && cc >= 0 &&
                    cc < grid[0].size() && grid[rr][cc] == '1' &&
                    visited.find(to_string(rr) + "," + to_string(cc)) ==
                        visited.end()) {
                    q.push({rr, cc});
                    visited.insert(to_string(rr) + "," + to_string(cc));
                }
            }
        }
    }
};

File: 200-number-of-islands/number-of-islands.go
func dfs(grid [][]byte, i, j int) {
    if (i < 0 || i >= len(grid)) || (j < 0 || j >= len(grid[0])) || grid[i][j] == '0' {
        return
    }

    grid[i][j] = '0'
    
    dfs(grid, i+1, j)
    dfs(grid, i-1, j)
    dfs(grid, i, j+1)
    dfs(grid, i, j-1)
}

func numIslands(grid [][]byte) int {
    cnt := 0
    for i := 0; i < len(grid); i++ {
        for j := 0; j < len(grid[0]); j++ {
            if grid[i][j] == '1' {
                cnt++
                dfs(grid, i, j)
            }
        }
    }
    return cnt
}